service/src/main/frontend/src/functions/formatters.ts

```diff
@@ -0,0 +1,107 @@
+/* eslint-disable no-var */
+import { Ref, WritableComputedRef, computed, defineComponent, ref } from 'vue';
+
+// Parses a Value to remove uncessessary 'undefined' or empty values.
+export function parseVal(colName: any, colValue: any): string {
+  if (colName === 'Types' || colName === 'Descriptions') {
+    if (colValue == undefined) {
+      return '';
+    } else {
+      return colValue.toString();
+    }
+  } else {
+    return colValue;
+  }
+}
+
+// Produces the max substring for the table, adds '...' if above 34 chars.
+export function maxSubstring(str: any, colName: any): any {
+  if (str == undefined) {
+    return;
+  } else if ((colName === 'fieldName' || colName === 'internalFieldName') && str.length > 32) {
+    return str.substring(0, 30) + ' ...';
+  } else if ((colName === 'Types') && str.length > 14) {
+    // Types is offset by 2 to prevent overlapping in 'Tokenized' Column
+    return str.substring(0, 9) + ' ...';
+  } else if ((colName === 'Descriptions') && str.length > 24) {
+    return str.substring(0, 22) + ' ...';
+  } else if ((colName === 'CopyPaste') && str.length > 42) {
+    return str.substring(0, 40) + ' ...';
+  } else {
+    return str;
+  }
+}
+
+// Defines how the expandability is parsed on the table.
+export function buttonParse(col: any, row: any): boolean {
+  return row.button == 1;
+}
+
+// Toggles how the row collapses based on the DOM. Filters visible rows.
+export function toggleVisibility(row: any) {
+  row.toggleVisibility();
+}
+
+// Set the Visibility in DOM, sorts and filters by lastUpdated, and the respective row to render button.
+export function setVisibility(rows: readonly any[]) {
+  const fieldVisibility: Map<string, Ref<boolean>> = new Map<
+    string,
+    Ref<boolean>
+  >();
+  const buttonValues: Map<string, number> = new Map<string, number>();
+
+  for (const row of rows) {
+    let maxUp: number = row.lastUpdated;
+    const fieldUp: any = row.internalFieldName;
+    for (const scan of rows) {
```

Comment:
Could you please clarify what you mean? `buttonValues` is a Map with `internalFieldName` and `leastRecentlyUpdated`. The if statement on lines 67-73 updates the button element in the DOM if `buttonValues` contains `fieldUp`. Any insights on simplifying this would be helpful!

Reply:
What caught my eye was the double for-loop over `rows`. I believe the overall objective for this block is to find the most recent row for the field. The inner for-loop seems to be running needlessly in the case we already found the most recent row for the current field.

Said another way, let's say `fieldUp` is `hamburger` on the first iteration of the outer for-loop and on the second it's also `hamburger`. We've already found the most recent row for `hamburger` so we can skip the inner for-loop.

Reply:
It's very possible I'm missing the intent of this code block. But usually looping over the same array twice within a single block indicates we could improve the algorithm.